packages/jsii-pacmak/lib/targets/python/type-name.ts

```diff
@@ -0,0 +1,354 @@
+  readonly #element: TypeName;
+
+  public constructor(element: TypeName) {
+    this.#element = element;
+  }
+
+  public pythonType(context: NamingContext) {
+    return `typing.Mapping[str,${this.#element.pythonType(context)}]`;
+  }
+
+  public requiredImports(context: NamingContext) {
+    return mergePythonImports(this.#element.requiredImports(context), { typing: new Set(['']) });
+  }
+}
+
+class List implements TypeName {
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #element: TypeName;
+
+  public constructor(element: TypeName) {
+    this.#element = element;
+  }
+
+  public pythonType(context: NamingContext) {
+    return `typing.List[${this.#element.pythonType(context)}]`;
+  }
+
+  public requiredImports(context: NamingContext) {
+    return mergePythonImports(this.#element.requiredImports(context), { typing: new Set(['']) });
+  }
+}
+
+class Optional implements TypeName {
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #wrapped: TypeName;
+
+  public constructor(wrapped: TypeName) {
+    this.#wrapped = wrapped;
+  }
+
+  public pythonType(context: NamingContext) {
+    const optionalType = this.#wrapped.pythonType({ ...context, ignoreOptional: true });
+    if (context.ignoreOptional || this.#wrapped === Primitive.ANY) {
+      return optionalType;
+    }
+    return `typing.Optional[${optionalType}]`;
+  }
+
+  public requiredImports(context: NamingContext) {
+    const wrapped = this.#wrapped.requiredImports({ ...context, ignoreOptional: true });
+    if (context.ignoreOptional) {
+      return wrapped;
+    }
+    return mergePythonImports(wrapped, { typing: new Set(['']) });
+  }
+}
+
+class Primitive implements TypeName {
+  private static readonly BOOL = new Primitive('bool', {});
+  private static readonly DATE = new Primitive('datetime.datetime', { 'datetime': new Set(['']) });
+  private static readonly JSII_NUMBER = new Primitive('jsii.Number', {}); // "jsii" is always already imported!
+  private static readonly STR = new Primitive('str', {});
+  private static readonly JSON = new Primitive('typing.Mapping[typing.Any, typing.Any]', { 'typing': new Set(['']) });
+
+  public static readonly ANY = new Primitive('typing.Any', { 'typing': new Set(['']) });
+  public static readonly NONE = new Primitive('None', {});
+
+  public static of(type: PrimitiveTypeReference): TypeName {
+    switch (type.primitive) {
+      case PrimitiveType.Boolean:
+        return Primitive.BOOL;
+      case PrimitiveType.Date:
+        return Primitive.DATE;
+      case PrimitiveType.Number:
+        return Primitive.JSII_NUMBER;
+      case PrimitiveType.String:
+        return Primitive.STR;
+      case PrimitiveType.Json:
+        return Primitive.JSON;
+      case PrimitiveType.Any:
+      default:
+        return Primitive.ANY;
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #pythonType: string;
```

Comment:
What's the benefit of using the new javascript private annotations over typescript's `private`? I would expect typescript to compile `private` to `#`, but otherwise from a readability and coding standards perspective it seems like it's not adding much value.

Reply:
TypeScript does *not* compile `private` to `#`. The semantic is different. Using `private foo`, I can always `as any` my way to access the field; where `#` actually makes access to the value impossible from outside of the defining type. It's better encapsulation and less risk of "inadvertently" overriding a private.